datumaro/components/media.py

```diff
@@ -0,0 +1,188 @@
+import os
+import os.path as osp
+import shutil
+
+import numpy as np
+
+from datumaro.util.image import (
+    _image_loading_errors, decode_image, lazy_image, save_image,
+)
+
+
+class MediaElement:
+    def __init__(self, path: str) -> None:
+        self._path = path
+
+    @property
+    def path(self) -> str:
+        """Path to the media file"""
+        return self._path
+
+    @property
+    def ext(self) -> str:
+        """Image file extension"""
+        return osp.splitext(osp.basename(self.path))[1]
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, __class__):
+            return False
+        return self._path == other._path
+
+class Image(MediaElement):
+    def __init__(self,
+            data: Union[np.ndarray, Callable[[str], np.ndarray], None] = None,
+            *,
+            path: Optional[str] = None,
+            size: Optional[Tuple[int, int]] = None):
+        assert size is None or len(size) == 2, size
+        if size is not None:
+            assert len(size) == 2 and 0 < size[0] and 0 < size[1], size
+            size = tuple(map(int, size))
+        self._size = size # (H, W)
+        if not self._size and isinstance(data, np.ndarray):
+            self._size = data.shape[:2]
+
+        assert path is None or isinstance(path, str), path
+        if path is None:
+            path = ''
+        elif path:
+            path = osp.abspath(path).replace('\\', '/')
+        self._path = path
+
+        if not isinstance(data, np.ndarray):
+            assert path or callable(data), "Image can not be empty"
+            assert data is None or callable(data)
+            if path and osp.isfile(path) or data:
+                data = lazy_image(path, loader=data)
+        self._data = data
+
+    @property
+    def data(self) -> np.ndarray:
+        """Image data in BGR HWC [0; 255] (float) format"""
+
+        if callable(self._data):
+            data = self._data()
+        else:
+            data = self._data
+
+        if self._size is None and data is not None:
+            self._size =  tuple(map(int, data.shape[:2]))
+        return data
+
+    @property
+    def has_data(self) -> bool:
+        return self._data is not None
+
+    @property
+    def has_size(self) -> bool:
+        return self._size is not None or isinstance(self._data, np.ndarray)
+
+    @property
+    def size(self) -> Optional[Tuple[int, int]]:
+        """Returns (H, W)"""
+
+        if self._size is None:
+            try:
+                data = self.data
+            except _image_loading_errors:
+                return None
+            if data is not None:
+                self._size = tuple(map(int, data.shape[:2]))
+        return self._size
+
+    def __eq__(self, other):
+        if isinstance(other, np.ndarray):
+            return self.has_data and np.array_equal(self.data, other)
+
+        if not isinstance(other, __class__):
+            return super().__eq__(other)
```

Comment:
Seems like this logic can lead to a violation of transitivity, for example:

```
a = Image(path=p, data=d1)
b = Image(path=p, data=d2)
c = MediaElement(path=p)

assert a == c
assert b == c
assert a != b
```

IMO, an instance of `MediaElement` should not compare equal to an instance of a subclass.